# 01 Importing Libraries

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [2]:
# creating path to products.csv

productpath = r'C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Original Data\products.csv'

In [3]:
# importing products data from original folder

df_prods = pd.read_csv(os.path.join(productpath), index_col = False)

In [4]:
# creating path to orders_wrangled.csv

orderpath = r"C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Prepared Data\orders_wrangled.csv"

In [5]:
# importing orders data from prepared folder

df_ords = pd.read_csv(os.path.join(orderpath), index_col = False)

# 03 Mixed Data

In [7]:
# creates a new dataframe called df_test

df_test = pd.DataFrame()

In [8]:
# creates a new column, mix, within df_test and fills it with numeric, string, and boolean values

df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Convert data type to string

df_test['mix'] = df_test['mix'].astype('str')

# 04 Missing Values

In [12]:
# Checking for missing values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Create new dataframe (df_nan) containing all null values from product_name column within df_prods

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Check number of rows and columns in df_prods

df_prods.shape

(49693, 5)

In [16]:
# create new dataframe, df_prods_clean without NaN values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Check number of rows and columns in df_prods_clean

df_prods_clean.shape

(49677, 5)

# 05 Duplicates

In [18]:
# Searching for duplicates in df_prods_clean and create a subset

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
# Checking output

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
# Check shape of df

df_prods_clean.shape

(49677, 5)

In [21]:
# Creating a new dataframe without the duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

# 4.5 Data Consistency Checks Task

## 1 & 2

In [23]:
# Check descriptive statistics for orders df

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [24]:
# Checking orders df

df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 8 columns]

#### 2) Looking at the count of days_since_prior_order, there appears to be multiple null values

## 3

In [25]:
# Checking for mixed types in df_ords dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [26]:
# Checking data types of different columns

df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#### 3) There are no columns containing mixed types of data

## 4

#### 4) N/A as no mixed data types

## 5

In [27]:
# Checking for missing values in df_ords

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### 5) There are 206,209 missing values in the days_since_prior_order column

#### 206,209 is the same number of unique user_id's so there is one value missing for each. This is presumably the first order that each user has placed

## 6

In [32]:
# Creating a new variable that acts as a flag based on missing values

df_ords['first_order'] = np.select([df_ords.days_since_prior_order.isnull()],[1])

In [31]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN            1  
1                        7                    15.0            0  
2                       12                    21.0            0  
3                        7                    29.0            0  
4                       15                    28.0            0  
...                    ...                     ...          ...  
3421078                 18                    29.0            0  
3421079                 10                    30.0            0  
3421080                 12                    18.0            0  
3421081                 12                     7.0            0  
3421082                 14                    30.0            0  

[3421083 rows x 9 columns]

#### 6) The missing values are still valuable to the data so I decided to create a new variable that would flag up whether it was the first order for that customer

## 7

In [34]:
# Running a check for duplicate values in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]

In [35]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, first_order]
Index: []

#### 7) There are no duplicates in the df_ords dataframe

## 8

#### 8) There were no duplicates to address

## 9

In [37]:
# Creating path to Prepared Data folder

path = r"C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Prepared Data"

In [38]:
# Exporting df_ords to IC Project Folder > Data

df_ords.to_csv(os.path.join(path, 'Orders_checked.csv'))

In [39]:
# Exporting df_prods_clean_no_dups to IC Project Folder > Data

df_prods_clean_no_dups.to_csv(os.path.join(path, 'Products_cleaned.csv'))